File: OI-ACM/Graph/ChainForwardStar.cpp
/**
 * @file    :   LinkEdge
 * @author  :   Tanphoon
 * @date    :   2023/02/28 16:33
 * @version :   2023/02/28 16:33
 */
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

struct ChainForwardStar {
    using Edge = pair<int, int>;
    vector<int> head, next;
    vector<Edge> edges;
    int E;
    ChainForwardStar(int V = 0) { head.resize(V + 1, -1); }
    void addEdge(int u, int v, int w = 1) {
        edges.push_back({v, w});
        next.push_back(head[u]);
        head[u] = E++;
    }
    vector<Edge> operator[](int u) {
        vector<Edge> res;
        for (int i = head[u]; ~i; i = next[i])
            res.push_back(edges[i]);
        return res;
    }
};
int main() {
    int V, E;
    cin >> V >> E;
    ChainForwardStar G(V);
    for (int i = 0; i < E; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        G.addEdge(u, v, w);
    }
    for (int u = 0; u < V; u++) {
        for (auto &[v, w] : G[u])
            cout << u << " " << v << " " << w << endl;
        cout << endl;
    }
}

File: OI-ACM/Graph/ChainForwardStar.py
class ChainForwardStar:

    def __init__(self, V) -> None:
        self.V = V + 1
        self.E = 0
        self.head = [-1] * (V + 1)
        self.next = []
        self.edges = []

    def addEdge(self, u, v, w=1):
        self.edges.append((v, w))
        self.next.append(self.head[u])
        self.head[u] = self.E
        self.E += 1

    def __getitem__(self, u):
        i = self.head[u]
        while i != -1:
            yield self.edges[i]
            i = self.next[i]


V = 6
E = 11
txt = [[1, 2, 24], [2, 1, 54], [5, 2, 34], [6, 3, 87], [2, 3, 124], [1, 4, 675], [2, 4, 345], [4, 1, 321], [2, 5, 587],
       [4, 5, 87], [5, 6, 956]]

Graph = ChainForwardStar(V)
for u, v, w in txt:
    Graph.addEdge(u, v, w)
for u in range(1, 7):
    for v, w in Graph[u]:
        print(u, v, w)
    print()